### PR TITLE
Backlight refactor

### PR DESCRIPTION
**Pull request by ReFil**
Code has been refactored to share a lot of features with RGB Underglow
Auto off now uses pointer to make code smaller, also maintains state before power off properly
implemented a set brightness behaviour, not because it was a useful feature but to allow for correct split locality work with only a slight change

### CODE REVIEW

**Comment by bortoz** [comment]
Thank you for your contribution!
As I said earlier, there are some conflicts that need to be resolved. To do that you have to rebase your branch to my last commit. If you are not familiar with git, I can explain how to do it.

**Comment by ReFil** [comment]
Apologies, finally got the rebasing stuff figured out

File: app/src/backlight.c [hunk 1]
```diff
@@ -42,56 +39,41 @@ static struct backlight_state state = {.brightness = CONFIG_ZMK_BACKLIGHT_BRT_ST
 
 static int zmk_backlight_update() {
     uint8_t brt = state.on ? state.brightness : 0;
-    for (int i = 0; i < BACKLIGHT_NUM_LEDS; i++) {
-        int rc = led_set_brightness(backlight_dev, i, brt);
```
Comment (bortoz) [inline app/src/backlight.c]: This section is required if you have multiple LEDs, see [https://deploy-preview-904--zmk.netlify.app/docs/features/backlight#multiple-backlight-leds](https://deploy-preview-904--zmk.netlify.app/docs/features/backlight#multiple-backlight-leds).
What is the reason for removing it?

File: app/src/backlight.c [hunk 2]
```diff
@@ -42,56 +39,41 @@ static struct backlight_state state = {.brightness = CONFIG_ZMK_BACKLIGHT_BRT_ST
...
-        int rc = read_cb(cb_arg, &state, sizeof(state));
-        if (rc < 0) {
-            return rc;
+        rc = read_cb(cb_arg, &state, sizeof(state));
+        if (rc >= 0) {
+            return 0;
         }
 
-        return zmk_backlight_update();
+        return rc;
```
Comment (bortoz) [inline app/src/backlight.c]: `zmk_backlight_update` must be called after the settings are loaded. Why did you change it?

File: app/src/backlight.c [hunk 3]
```diff
@@ -42,56 +39,41 @@ static struct backlight_state state = {.brightness = CONFIG_ZMK_BACKLIGHT_BRT_ST
...
-#if IS_ENABLED(CONFIG_SETTINGS)
-    k_delayed_work_cancel(&backlight_save_work);
-    return k_delayed_work_submit(&backlight_save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
-#else
-    return 0;
 #endif
-}
 
-static int zmk_backlight_init(const struct device *_arg) {
+int zmk_backlight_init(const struct device *_arg) {
```
Comment (bortoz) [inline app/src/backlight.c]: This function should be declared as static.

File: app/src/backlight.c [hunk 4]
```diff
@@ -42,56 +39,41 @@ static struct backlight_state state = {.brightness = CONFIG_ZMK_BACKLIGHT_BRT_ST
...
-        return zmk_backlight_update();
+        return rc;
     }
 
     return -ENOENT;
 }
 
-static struct settings_handler backlight_conf = {.name = "backlight",
-                                                 .h_set = backlight_settings_set};
+struct settings_handler backlight_conf = {.name = "backlight", .h_set = backlight_settings_set};
```
Comment (bortoz) [inline app/src/backlight.c]: This variable should be declared as static.

File: app/src/backlight.c [hunk 5]
```diff
@@ -114,11 +96,31 @@ static int zmk_backlight_init(const struct device *_arg) {
...
+    k_delayed_work_cancel(&backlight_save_work);
+    return k_delayed_work_submit(&backlight_save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
+#else
+    return 0;
+#endif
+}
+
+int zmk_backlight_get_state(bool *on_off) {
+    if (!backlight_dev)
+        return -ENODEV;
```
Comment (bortoz) [inline app/src/backlight.c]: `backlight_dev` is always not `NULL`, this check is useless.

File: app/src/behaviors/behavior_backlight.c [hunk 6]
```diff
@@ -20,21 +20,59 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
 static int behavior_backlight_init(const struct device *dev) { return 0; }
 
+static int
+on_keymap_binding_convert_central_state_dependent_params(struct zmk_behavior_binding *binding,
+                                                         struct zmk_behavior_binding_event event) {
```
Comment (bortoz) [inline app/src/behaviors/behavior_backlight.c:25]: I see no reason to add this function. You can do all the checks inside `on_keymap_binding_pressed.` Adding this function only increases the complexity and readability of the code.

**Review by bortoz — CHANGES_REQUESTED** [review]
Many of the changes made are useless or wrong. Backlight and underglow are two different features and trying to make them similar only worsens the quality of the code.

File: app/src/behaviors/behavior_backlight.c (hunk 6 above)
Comment (ReFil) [reply app/src/behaviors/behavior_backlight.c:25]: This is for the split locality support. When a key is pressed on the peripheral it synchronises the two sides. As is the code can end up in a state where one side is on and the other is off and pressing the toggle key will turn one side on and the other off, and the brightness will be out of sync

File: app/src/backlight.c (hunk 5 above)
Comment (ReFil) [reply app/src/backlight.c]: This is if backlight behaviours are called without backlight being defined

File: app/src/backlight.c (hunk 1 above)
Comment (ReFil) [reply app/src/backlight.c]: Ah I see i didn't understand the need for multiple LEDS when it's all driven through one channel, will add back

File: app/src/backlight.c (hunk 2 above)
Comment (ReFil) [reply app/src/backlight.c]: settings_set doesn't load the settings thus it doesn't need to update the lights, it saves the current settings. The settings are loaded in backlight_init

File: app/src/backlight.c (hunk 4 above)
Comment (ReFil) [reply app/src/backlight.c]: Will fix

File: app/src/backlight.c (hunk 2 above)
Comment (bortoz) [reply app/src/backlight.c]: According to the documentation ([https://docs.zephyrproject.org/latest/reference/settings/index.html#handlers](https://docs.zephyrproject.org/latest/reference/settings/index.html#handlers)), `h_set` is called "_when the value is loaded from persisted storage_", the settings are saved using `settings_save_one`.

File: app/src/backlight.c (hunk 5 above)
Comment (bortoz) [reply app/src/backlight.c]: If the device is not defined, a compilation error is generated, otherwise the device is always non-`NULL`.

File: app/src/behaviors/behavior_backlight.c (hunk 6 above)
Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c:25]: I understand, my bad

File: app/src/backlight.c (hunk 5 above)
Comment (ReFil) [reply app/src/backlight.c]: Using a split however you could have it properly defined on one side calling the function on the other

Comment (bortoz) [reply app/src/backlight.c]: I think this shouldn't be allowed and I don't even think it will work as it is not possible to enable `CONFIG_ZMK_BACKLIGHT` without a device

File: app/src/backlight.c (hunk 2 above)
Comment (ReFil) [reply app/src/backlight.c]: I was under the impression that the loading update would be caught in the backlight init's call of zmk_backlight_update. Looking again I see it submits a delayed work request so it might get missed by that call. putting it in here too is correct, sorry, will fix

Comment (ReFil) [reply app/src/backlight.c]: In my testing the settinigs load call was executed quickly enough that the update in the init worked every time however given more complicated features in the future definitely worth keeping. I was trying to minimise spurious function calls

File: app/src/backlight.c (hunk 5 above)
Comment (ReFil) [reply app/src/backlight.c]: but you can include the behaviours file and have the central attempt to call the function on the peripheral. The ability to ensure that edge cases get caught seems like a worthwhile addition? I can remove it if you're sure it's superfluous

Comment (bortoz) [reply app/src/backlight.c]: You can include the behavior header, however, unless you enable `CONFIG_ZMK_BACKLIGHT`, any presses will be ignored.

Comment (ReFil) [reply app/src/backlight.c]: Ok i see, will remove

File: app/src/backlight.c [hunk 7]
```diff
@@ -22,16 +22,13 @@
...
 BUILD_ASSERT(DT_HAS_CHOSEN(zmk_backlight),
              "CONFIG_ZMK_BACKLIGHT is enabled but no zmk,backlight chosen node found");
 
 static const struct device *const backlight_dev = DEVICE_DT_GET(DT_CHOSEN(zmk_backlight));
 
-#define CHILD_COUNT(...) +1
-#define DT_NUM_CHILD(node_id) (DT_FOREACH_CHILD(node_id, CHILD_COUNT))
-
-#define BACKLIGHT_NUM_LEDS (DT_NUM_CHILD(DT_CHOSEN(zmk_backlight)))
-
```
Comment (bortoz) [inline app/src/backlight.c:34]: These macros should be included again.

File: app/src/backlight.c [hunk 8]
```diff
@@ -55,15 +52,16 @@ static int zmk_backlight_update() {
...
         if (len != sizeof(state)) {
             return -EINVAL;
         }
 
-        int rc = read_cb(cb_arg, &state, sizeof(state));
-        if (rc < 0) {
-            return rc;
+        rc = read_cb(cb_arg, &state, sizeof(state));
+        if (rc >= 0) {
+            return 0;
```
Comment (bortoz) [inline app/src/backlight.c]: This part should be fixed.

File: app/Kconfig [hunk 9]
```diff
@@ -542,4 +542,3 @@ osource "$(ZMK_CONFIG)/boards/shields/*/Kconfig.shield"
 
 
 source "Kconfig.zephyr"
-
```
Comment (bortoz) [inline app/Kconfig]: Minor: can you revert this change?

File: app/src/backlight.c [hunk 10]
```diff
@@ -142,23 +143,42 @@ int zmk_backlight_set_brt(int brt) {
...
 
-int zmk_backlight_get_brt() { return state.on ? state.brightness : 0; }
+int zmk_backlight_toggle() { return state.on ? zmk_backlight_off() : zmk_backlight_on(); }
 
-int zmk_backlight_toggle() { return zmk_backlight_set_on(!state.on); }
-
-int zmk_backlight_on() { return zmk_backlight_set_on(true); }
+int zmk_backlight_set_brt(uint8_t brightness) {
+    if (brightness > BRT_MAX) {
+        return -ENOTSUP;
```
Comment (bortoz) [inline app/src/backlight.c]: I think this function shouldn't return an error, consider the following situation:
- `state.brightness = 100`
- `state.on = false`
- press `BL_INC`

Expected behavior: turn on the backlight.
Actual behavior: an error is returned.

File: app/src/behaviors/behavior_backlight.c [hunk 11]
```diff
@@ -20,21 +20,59 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
...
+        if (err) {
+            LOG_ERR("Failed to get backlight state (err %d)", err);
+            return err;
+        }
+
+        binding->param1 = state ? BL_OFF_CMD : BL_ON_CMD;
+        break;
+    }
+    case BL_INC_CMD: {
+        uint8_t brightness = zmk_backlight_calc_brt(1);
```
Comment (bortoz) [inline app/src/behaviors/behavior_backlight.c:32]: This logic doesn't work, check this comment: https://github.com/zmkfirmware/zmk/pull/904#discussion_r689291345

**Review by bortoz — CHANGES_REQUESTED** [review]
A few more comments

File: app/src/behaviors/behavior_backlight.c (hunk 11 above)
Comment (ReFil) [reply app/src/behaviors/behavior_backlight.c:32]: It worked on my board when I was testing it

File: app/src/backlight.c (hunk 10 above)
Comment (ReFil) [reply app/src/backlight.c]: Changed it to set the on state properly. The set brightness function is meant primarily for syncing between both sides of a split

File: app/Kconfig (hunk 9 above)
Comment (ReFil) [reply app/Kconfig]: Fixed, apologies. Atom automatically removed it

File: app/src/backlight.c (hunk 8 above)
Comment (ReFil) [reply app/src/backlight.c]: Reverted

File: app/src/backlight.c (hunk 7 above)
Comment (ReFil) [reply app/src/backlight.c:34]: Included

File: app/src/behaviors/behavior_backlight.c (hunk 11 above)
Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c:32]: It works most of the time but there are some edge cases, there are many examples in the original PR comments. I'll try to add some tests to detect those corner cases.

Comment (ReFil) [reply app/src/behaviors/behavior_backlight.c:32]: Looking through that comment I cannot see what edge cases mentioned won't get caught within my Inc and Dec functions. It's very close to the currently implemented RGB underglow behaviours API

File: app/src/backlight.c (hunk 10 above)
Comment (bortoz) [reply app/src/backlight.c]: The new change also doesn't work, if `state.on = false`, you can set the brightness to any value greater than `BRT_MAX`.

File: app/src/behaviors/behavior_backlight.c (hunk 11 above)
Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c:32]: I'm sorry, I misunderstood your code. Your code works fine but you have a different interpretation than mine.
I assume that pressing `BL_INC` when `state.on = false`, turns on the backlight, check my code: [https://github.com/bortoz/zmk/blob/backlight/app/src/backlight.c#L154](https://github.com/bortoz/zmk/blob/backlight/app/src/backlight.c#L154). There was also some thought about that in the discussion I have liked.
I honestly don't remember why I made that choice, however I think this implementation is fine too.

Comment (ReFil) [reply app/src/behaviors/behavior_backlight.c:32]: yes pressing BL_INC when it's off turns on the backlight, the `zmk_backlight_calc_brt` function limits the backlight to within the range of the max value, and `state.on` is calculated automatically

File: app/src/backlight.c (hunk 10 above)
Comment (ReFil) [reply app/src/backlight.c]: Because the `set_brt` function works in conjunction with the `calc_brt` function which automatically auto ranges to within the max min values when using the `BL_INC` the only reason the `set_brt` function would receive an out of range brightness  is if the user configures their keymap to do so, I'm not sure the best way to deal with this is just clamp the brightness regardless but I can change it

File: app/include/dt-bindings/zmk/backlight.h [hunk 12]
```diff
@@ -4,8 +4,16 @@
...
+#define BL_INC_CMD 3
+#define BL_DEC_CMD 4
+#define BL_SET_CMD 5
+
+#define BL_TOG BL_TOG_CMD 0
+#define BL_ON BL_ON_CMD 0
+#define BL_OFF BL_OFF_CMD 0
+#define BL_INC BL_INC_CMD 0
+#define BL_DEC BL_DEC_CMD 0
+#define BL_SET BL_SET_CMD BRIGHTNESS
```
Comment (bortoz) [inline app/include/dt-bindings/zmk/backlight.h]: `BRIGHTNESS` is not defined, you may want to remove it.

File: app/src/backlight.c [hunk 13]
```diff
@@ -22,6 +22,8 @@
 
 LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
+#define BRT_MAX 100
```
Comment (bortoz) [inline app/src/backlight.c]: Minor: can you move this macro near to the others?

File: app/src/backlight.c [hunk 14]
```diff
@@ -142,23 +147,44 @@ int zmk_backlight_set_brt(int brt) {
...
+uint8_t zmk_backlight_calc_brt(int direction) {
+    uint8_t brightness = state.brightness;
+
+    int b = brightness + (direction * CONFIG_ZMK_BACKLIGHT_BRT_STEP);
+    brightness = CLAMP(b, 0, BRT_MAX);
+
+    return brightness;
+}
+
+int zmk_backlight_change_brt(int direction) {
```
Comment (bortoz) [inline app/src/backlight.c]: See https://github.com/zmkfirmware/zmk/pull/904#discussion_r688176294

> Maybe use "adjust" instead of "change" to better indicate that this changes relative to the existing value, not changes to a new value.

File: app/src/backlight.c [hunk 15]
```diff
@@ -142,23 +147,44 @@ int zmk_backlight_set_brt(int brt) {
...
 
-int zmk_backlight_dec() {
-    return zmk_backlight_set_brt(state.brightness - CONFIG_ZMK_BACKLIGHT_BRT_STEP);
+uint8_t zmk_backlight_calc_brt(int direction) {
+    uint8_t brightness = state.brightness;
+
+    int b = brightness + (direction * CONFIG_ZMK_BACKLIGHT_BRT_STEP);
+    brightness = CLAMP(b, 0, BRT_MAX);
+
+    return brightness;
```
Comment (bortoz) [inline app/src/backlight.c]: This function can be simplified to something like this:

```c
int b = state.brightness + (direction * CONFIG_ZMK_BACKLIGHT_BRT_STEP);
return CLAMP(b, 0, BRT_MAX);
```

File: app/src/behaviors/behavior_backlight.c [hunk 16]
```diff
@@ -20,21 +20,59 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
...
 
+static int
+on_keymap_binding_convert_central_state_dependent_params(struct zmk_behavior_binding *binding,
+                                                         struct zmk_behavior_binding_event event) {
+    switch (binding->param1) {
+    case BL_TOG_CMD: {
+        bool state;
+        int err = zmk_backlight_get_state(&state);
+        if (err) {
+            LOG_ERR("Failed to get backlight state (err %d)", err);
```
Comment (bortoz) [inline app/src/behaviors/behavior_backlight.c]: The function `zmk_backlight_get_state` never returns an error so this check may be removed. Maybe it is also worth using the old signature:  `bool zmk_backlight_is_on();`.

File: app/include/zmk/backlight.h [hunk 17]
```diff
@@ -6,14 +6,10 @@
 
 #pragma once
 
-int zmk_backlight_set_on(bool on);
-bool zmk_backlight_is_on();
-
-int zmk_backlight_set_brt(int brt);
-int zmk_backlight_get_brt();
```
Comment (bortoz) [inline app/include/zmk/backlight.h:13]: I think `zmk_backlight_get_brt` was useful. Why did you remove it?

**Review by bortoz — CHANGES_REQUESTED** [review]
Almost ready!

File: app/include/zmk/backlight.h (hunk 17 above)
Comment (ReFil) [reply app/include/zmk/backlight.h:13]: `calc_brt` with direction 0 gives the same result thus its not needed

Comment (bortoz) [reply app/include/zmk/backlight.h:13]: See https://github.com/bortoz/zmk/blob/backlight/app/src/backlight.c#L145.
They are not the same function, `zmk_backlight_get_brt` should return 0 if `state.on = false`. It is still possible to calculate this value but I think it is convenient to have this function.

Comment (ReFil) [reply app/include/zmk/backlight.h:13]: Fair enough, I will implement it

File: app/src/behaviors/behavior_backlight.c (hunk 16 above)
Comment (ReFil) [reply app/src/behaviors/behavior_backlight.c]: Ok will sort, not aure about `backlight_is_on` as opposed to get state but will change if you feel strongly about it

Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c]: If the function never returns an error, it is better to return the value directly, for example:
https://github.com/zmkfirmware/zmk/blob/main/app/include/zmk/usb.h#L24
https://github.com/zmkfirmware/zmk/blob/main/app/include/zmk/wpm.h#L9
https://github.com/zmkfirmware/zmk/blob/main/app/include/zmk/battery.h#L9

Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c]: As for the name, I think `get_state` is a bit ambiguous since in the code `state` refers to the whole state (brightness and on/off).

**Comment by ReFil** [comment]
Sorry this fix took so long i was away for a few days

File: app/src/behaviors/behavior_backlight.c (hunk 16 above)
Comment (bortoz) [reply app/src/behaviors/behavior_backlight.c]: The name `get_on` is ok, but I still think it should return the value directly in the same way as `get_brt`

**Review by bortoz — APPROVED** [review]
Looks good to me, I'll test on my board for a few days and then merge it.
